src/vmm/src/resources.rs

```diff
@@ -153,6 +159,67 @@ impl VmResources {
         Ok(resources)
     }
 
+    /// Configures VmResources from the state of the Vmm.
+    pub fn from_vmm(vmm: &Vmm) -> std::result::Result<Self, Error> {
+        let mut resources = Self::default();
+        resources.boot_config = None;
+
+        if let Ok(balloon_config) = vmm.balloon_config() {
+            resources
+                .set_balloon_device(BalloonDeviceConfig::from(balloon_config))
+                .map_err(Error::BalloonDevice)?;
+        }
+
+        vmm.mmio_device_manager.for_each_virtio_device(|virtio_type, _device_id, _device_info, virtio_device| {
+            if virtio_type != TYPE_BLOCK {
+                return Ok(());
+            }
+
+            resources.set_block_device(BlockDeviceConfig::from(
+                virtio_device
+                    .lock()
+                    .expect("Poisoned lock")
+                    .as_any()
+                    .downcast_ref::<Block>()
+                    .unwrap()
+            ))
+            .map_err(Error::BlockDevice)
+        })?;
+
+        resources.set_vm_config(&VmConfig {
+            vcpu_count: Some(vmm.vcpus_handles.len() as u8),
+            mem_size_mib: Some(mem_size_mib(vmm.guest_memory()) as usize),
+            ht_enabled: Some(false), // TODO: how to determine if HT enabled?
+            cpu_template: None, // TODO: how to determine cpu_template?
+            track_dirty_pages: vmm
+                .guest_memory()
+                .iter()
+                .any(|region| (region.flags() as u32) & KVM_MEM_LOG_DIRTY_PAGES != 0)
+        }).map_err(Error::VmConfig)?;
+
+        resources.net_builder = NetBuilder::new();
+        vmm.mmio_device_manager.for_each_virtio_device(|virtio_type, _device_id, _device_info, virtio_device| {
+            if virtio_type != TYPE_NET {
+                return Ok(());
+            }
+
+            resources.build_net_device(NetworkInterfaceConfig::from(
+                virtio_device
+                    .lock()
+                    .expect("Poisoned lock")
+                    .as_any()
+                    .downcast_ref::<Net>()
+                    .unwrap()
+            )).map_err(Error::NetDevice)?;
+
+            Ok(())
+        })?;
+
+        // TODO: how to determine vsock & mmds config (seems like should be able to use for_each_virtio_device
```

Comment:
I also noticed that currently (even if not resumed from a snapshot), the outputted MMDS config is `None` if a custom IP address has not been specified but a net device was configured with `allow_mmds_requests: true`. We should actually output the default MMDS config in this case.
You could include this fix in the PR.

Reply:
> the outputted MMDS config is None if a custom IP address has not been specified but a net device was configured with allow_mmds_requests: true. We should actually output the default MMDS config in this case.

I'm a little confused by this; could you elaborate? I see the mmds-config appears to be always be null when hitting the `GET /vm/config` endpoint which I guess is what you're saying is a problem. Specifically the part I'm confused on is what you mean by "if a custom IP address has not been specified but a net device was configured with allow_mmds_requests: true". I'm reading through the MMDS documentation again to see if I can make more sense of what you mean by that.

Reply:
> I'm a little confused by this; could you elaborate? I see the mmds-config appears to be always be null when hitting the GET /vm/config endpoint which I guess is what you're saying is a problem. Specifically the part I'm confused on is what you mean by "if a custom IP address has not been specified but a net device was configured with allow_mmds_requests: true". I'm reading through the MMDS documentation again to see if I can make more sense of what you mean by that.

If a user does not customise the default IP address used for MMDS, there is no MMDS IP address outputted when dumping the vm config (while there is actually a default one).

We could open a separate issue for this, you don't need to tackle it in this PR. Let's take them one step at a time. WDYT?

Reply:
This will be solved when implementing the proposal from this comment: https://github.com/firecracker-microvm/firecracker/pull/2806#discussion_r812532289